mq/main/comm-util/src/test/java/com/sun/messaging/jmq/util/lists/FifoSetTest.java

```diff
@@ -0,0 +1,109 @@
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * package com.sun.messaging.jmq.util;
+ */
+package com.sun.messaging.jmq.util.lists;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for usage of FifoSet and its subclasses
+ *
+ * @author jonathan coustick
+ * @see FifoSet
+ * @see PriorityFifoSet
+ * @see NFLPriorityFifoSet
+ */
+public class FifoSetTest {
+
+    private static final String FIRST = "first";
+    private static final String SECOND = "second";
+    private static final String THIRD = "third";
+
+    @Test
+    public void fifoTest() {
+        FifoSet<String> testSet = new FifoSet<>();
+        Assertions.assertTrue(testSet.isEmpty());
+        testSet.add(FIRST);
+        testSet.add(SECOND);
+        testSet.add(THIRD);
+        Assertions.assertEquals(3, testSet.size());
+        Assertions.assertTrue(testSet.contains(SECOND));
+        Assertions.assertEquals(FIRST, testSet.first());
+        Assertions.assertEquals(FIRST, testSet.first());//not removed
+        Assertions.assertEquals(THIRD, testSet.last());
+        
+        testSet.remove(FIRST);
+        Assertions.assertEquals(SECOND, testSet.first());
+        Assertions.assertEquals(2, testSet.size());
+        testSet.clear();
+        Assertions.assertTrue(testSet.isEmpty());
+
+    }
+
+    @Test
+    public void PriorityFifoSet() {
+        PriorityFifoSet<String> testSet = new PriorityFifoSet<>();
+        Assertions.assertTrue(testSet.isEmpty());
+        testSet.add(2, FIRST);
+        testSet.add(SECOND);
+        testSet.add(0, THIRD);
+        Assertions.assertEquals(3, testSet.size());
+        Assertions.assertTrue(testSet.contains(SECOND));
+        Assertions.assertEquals(THIRD, testSet.first());
+        Assertions.assertEquals(THIRD, testSet.first());//not removed
+        Assertions.assertEquals(SECOND, testSet.last());
+        
+        testSet.remove(FIRST);
+        Assertions.assertEquals(THIRD, testSet.first());
+        Assertions.assertEquals(2, testSet.size());
+        testSet.clear();
+        Assertions.assertTrue(testSet.isEmpty());
+
+    }
+
+    @Test
+    public void NFLPriorityFifoSet() {
+        NFLPriorityFifoSet<String> testSet = new NFLPriorityFifoSet<>();
+        Assertions.assertTrue(testSet.isEmpty());
+        testSet.add(FIRST);
+        testSet.add(SECOND);
+        testSet.add(THIRD);
+        Assertions.assertEquals(3, testSet.size());
+        try {
+            Assertions.assertTrue(testSet.contains(SECOND));
+            Assertions.fail("Lock should havce been set");
+        } catch (AssertionError e) {
```

Comment:
What's tested here? Which assertion is silently consumed here?

Reply:
I don't understand - is `Assertions.assertTrue(testSet.contains(SECOND));` expected to fail, and if it doesn't - test shall fail?

Reply:
I might be wrong, but for me this `try-catch`
- can be removed
- or has to be reworked

because:
1. `testSet.contains` doesn't throw itself
2.
   1. assuming `contains` returns `false` - which I suppose is expected outcome - `assertTrue` does not hold and throws `AssertionError` which is caught and silently ignored 
   1. assuming `contains` returns `true` - which I guess is goal to be detected as failure here -  `assertTrue` holds and flow continues to `fail` - which throws `AssertionError` which is caught and silently ignored again
1. so regardless of the outcome of `contains(SECOND)` - nothing happens.

Do I misunderstand this section?

Reply:
I'm still not sure. Why not replace this try-catch with
```java
Assertions.assertThrows(AssertionError.class, () -> testSet.contains(SECOND));
```
then? It WILL fail if contains doesn't throw, while the current try-catch will pass anyway.

Reply:
testSet.contains is expected to throw an AssertionError. See:
https://github.com/eclipse-ee4j/openmq/blob/4af0b1313ab8c76430573e3cf4c097887aac56ba/mq/main/comm-util/src/main/java/com/sun/messaging/jmq/util/lists/FifoSet.java#L77
This lock is set in the constructor of NFLPriorityFifoSet:
https://github.com/eclipse-ee4j/openmq/blob/4af0b1313ab8c76430573e3cf4c097887aac56ba/mq/main/comm-util/src/main/java/com/sun/messaging/jmq/util/lists/NFLPriorityFifoSet.java#L75
and so if the thread does not hold the lock then an assertion error should be thrown, that is what the test is for.

Reply:
Should this `AE` be not thrown from `contains` - `Assertions.fail` is expected to fail the test?

Reply:
Thank you for your patience and provided explanations.

> > Should this `AE` be not thrown from `contains` - `Assertions.fail` is expected to fail the test?

> That's right

I though so (https://github.com/eclipse-ee4j/openmq/pull/573#discussion_r490340405).

Then - does it?

With _broken_ `contains` [by removing `assert`](https://github.com/eclipse-ee4j/openmq/pull/583/commits/74d0c0bdfc803c6bb3748581d7f158e3946ff937), test [passes](https://travis-ci.com/github/eclipse-ee4j/openmq/builds/185227584#L414) anyway. 
What does it mean to fail with `.fail`?
https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#fail(java.lang.String)
> Since this method throws an AssertionFailedError before its return statement...

and here's what happens;
```
org.opentest4j.AssertionFailedError: Lock should have been set
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:109)
	at com.sun.messaging.jmq.util.lists.FifoSetTest.NFLPriorityFifoSetTest(FifoSetTest.java:87)
```


https://ota4j-team.github.io/opentest4j/docs/1.2.0/api/org/opentest4j/AssertionFailedError.html:
```
     java.lang.Object
        java.lang.Throwable
            java.lang.Error
                java.lang.AssertionError
                    org.opentest4j.AssertionFailedError 
```
which is handled silently.

Reply:
*bangs head on wall*
Assertions.fail() throws an Assertion error, which is caught on the next line, so making the test pass. I've published a new commit which stores it as boolean, and then in a finally block checks it. A failure here means it is actually thrown and the test fails.

Reply:
I'm still not sure. Why not replace this try-catch with
```java
Assertions.assertThrows(AssertionError.class, () -> testSet.contains(SECOND));
```
then? It WILL fail if contains doesn't throw, while the current try-catch will pass anyway.

Reply:
Done.

Reply:
> Done.

Cool. I hope it's *done* because you like it :smile: 

Perhaps `OutOfLimitsException` in `HashMapTest` could be asserted in similar way...?

Reply:
Also done. My issue is that I don't always think of using functional interfaces when they are available, and I'm not used to Junit 5. Now I know they're there I will try to make more use of them in future.